### ISSUE TITLE
Does `write` have the right type?

### ISSUE THREAD

**Issue by hdgarrood** [purescript-node/purescript-node-streams#3]
Currently `write` is:

```
write :: forall r eff a. Writable r eff String -> a -> Eff eff Unit -> Eff eff Boolean
```

Should it be:

```
write :: forall r eff a. Writable r eff a -> a -> Eff eff Unit -> Eff eff Boolean
```
?

**Comment by hdgarrood** [comment]
Also what is the third argument for? Would be nice to explain this in the docs.

**Comment by hdgarrood** [comment]
Oh of course, it's a callback.

**Comment by garyb** [comment]
Your suggested change seems sensible to me, but then I'm surprised having `a` in the types here at all is safe, I'd have thought we'd need some class constraint or foreign type to represent things that streams can deal with.

**Comment by paf31** [comment]
I think the change is correct. The tricky thing is that Node stream types can change by calling things like `setEncoding`, so the current API is a bit of an approximation. Maybe indexed monads are the way to go, but I'm really not sure.

**Comment by hdgarrood** [comment]
The mutating nature of `setEncoding` is not very PureScript-y, so perhaps trying to provide a principled, typed interface to it in PureScript is the wrong approach.

As far as I can tell, `write` can only accept Strings or Buffers, but a given stream can accept either at any point; it's fine to `write` a String, then a Buffer, then another String. What about:

* Remove the last type parameter of `Stream`
* Have separate `writeBuffer` and `writeString` functions, where both work on any writable stream?

Reading is a little harder. I think we could deal with the fact that we don't know what type we might get back by inspecting the object in the FFI, and converting if necessary.

**Comment by hdgarrood** [comment]
My proposition is essentially: the fact that stream APIs can take or return either of `Buffer` or `String` seems to only be there for convenience, and I think we may be better off trying not to replicate that here.

**Comment by hdgarrood** [comment]
I tried to implement this but unfortunately there doesn't appear to be a way to ignore the encoding once one has been set. :( Indexed monads might work (I don't really know), but I think that's a bit fancy for a low-level API. I'd rather do one of the following:

* Throw an error inside `onData` we get anything other than a `Buffer`
* Change the callback type in `onData` to `Either String Buffer -> Eff eff Unit`

It seems that some Node.js libraries have opted for the first option in this situation, for example: https://github.com/senchalabs/connect/issues/966

**Comment by hdgarrood** [comment]
Fixed by #4
